tests/attention/test_vsa_block_sparse.py

```
@@ -0,0 +1,827 @@
+    cos_sim = _cosine_sim(o_vsa, o_dense)
+    print(
+        f"\n[accuracy] seqlen={seqlen:>6}  density={actual_density:.3f}"
+        f"  MAE={mae:.5f}  CosSim={cos_sim:.4f}"
+    )
+
+
+def test_vsa_performance_vs_dense():
+    """Performance table: full fastvideo VSA (compress + select) vs dense FlashInfer prefill.
+
+    VSA timing is broken into two parts:
+      compress_ms – Stage 1: compress attention (pool Q/K/V, dense attn on pooled tokens,
```

Ruff (0.15.12)

[warning] Docstring contains ambiguous `–` (EN DASH). Did you mean `-` (HYPHEN-MINUS)?

(RUF002)